core/quorum/api.go

```diff
@@ -0,0 +1,804 @@
+	ac := types.AcctInfoMap.GetAccount(account)
+
+	if ac == nil {
+		return ErrAccountNotThere, errors.New("account not there")
+	}
+
+	if ac.IsOrgAdmin && (ac.RoleId == s.permConfig.NwAdminRole || ac.RoleId == s.permConfig.OrgAdminRole) && (op == 1 || op == 3) {
+		return ErrOpNotAllowed, errors.New("operation not allowed on org admin account")
+	}
+
+	if ac.OrgId != orgId {
+		return ErrOrgNotOwner, errors.New("account does not belong to the organization passed")
+	}
+
+	if ac.Status == types.AcctBlacklisted {
+		return ErrBlacklistedAccount, errors.New("blacklisted account. operation not allowed")
+	}
+
+	// validate the op and node status and check if the op can be performed
+	if op != 1 && op != 2 && op != 3 {
+		return ErrOpNotAllowed, errors.New("invalid account status change operation")
+	}
+
+	if (op == 1 && ac.Status != types.AcctActive) || (op == 2 && ac.Status != types.AcctSuspended) {
+		return ErrOpNotAllowed, errors.New("account status change cannot be performed")
+	}
+	return ExecSuccess, nil
+}
+
+func (s *QuorumControlsAPI) checkOrgAdminExists(orgId, roleId string, account common.Address) (ExecStatus, error) {
+	ac := types.AcctInfoMap.GetAccount(account)
+
+	if ac != nil {
+		if ac.OrgId != orgId {
+			return ErrAccountInUse, errors.New("account part of another org")
+		}
+		if roleId != "" && roleId == s.permConfig.OrgAdminRole && ac.IsOrgAdmin {
+			return ErrAccountOrgAdmin, errors.New("account already org admin for the org")
+		}
+	}
+	return ExecSuccess, nil
+}
+
+func (s *QuorumControlsAPI) valSubOrgBreadthDepth(porgId string) (ExecStatus, error) {
+	org := types.OrgInfoMap.GetOrg(porgId)
+
+	if s.permConfig.SubOrgDepth.Cmp(org.Level) == 0 {
+		return ErrMaxDepth, errors.New("max depth for sub orgs reached")
+	}
+
+	if s.permConfig.SubOrgBreadth.Cmp(big.NewInt(int64(len(org.SubOrgList)))) == 0 {
+		return ErrMaxBreadth, errors.New("max breadth for sub orgs reached")
+	}
+
+	return ExecSuccess, nil
+}
+
+func (s *QuorumControlsAPI) checkNodeExists(url, enodeId string) bool {
+	node := types.NodeInfoMap.GetNodeByUrl(url)
+	if node != nil {
+		return true
+	}
+	// check if the same nodeid is in use with different port numbers
+	nodeList := types.NodeInfoMap.GetNodeList()
+	for _, n := range nodeList {
+		if enodeDet, er := enode.ParseV4(n.Url); er == nil {
+			if enodeDet.EnodeID() == enodeId {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func (s *QuorumControlsAPI) valNodeDetails(url string) (ExecStatus, error) {
+	// validate node id and
+	if len(url) != 0 {
+		enodeDet, err := enode.ParseV4(url)
+		if err != nil {
+			return ErrInvalidNode, errors.New("invalid node id")
+		}
+		// check if node already there
+		if s.checkNodeExists(url, enodeDet.EnodeID()) {
+			return ErrNodePresent, errors.New("duplicate node")
+		}
+	}
+	return ExecSuccess, nil
+}
+
+// executePermAction helps to execute an action in permission contract
+func (s *QuorumControlsAPI) executePermAction(action PermAction, args txArgs) ExecStatus {
```

Comment:
This function is super long. Maybe command pattern would help here

Reply:
rafactored code. Though have not used command patter. Want to test bit more before I push the changes.

Reply:
changes pushed